fuse/src/main/java/tachyon/fuse/TachyonFuseFs.java

```diff
@@ -0,0 +1,577 @@
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package tachyon.fuse;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import com.google.common.base.Preconditions;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+
+import tachyon.Constants;
+import tachyon.TachyonURI;
+import tachyon.client.file.TachyonFile;
+import tachyon.client.file.TachyonFileSystem;
+import tachyon.exception.FileAlreadyExistsException;
+import tachyon.exception.FileDoesNotExistException;
+import tachyon.exception.InvalidPathException;
+import tachyon.exception.TachyonException;
+import tachyon.thrift.FileInfo;
+
+import jnr.ffi.Pointer;
+import jnr.ffi.types.mode_t;
+import jnr.ffi.types.off_t;
+import jnr.ffi.types.size_t;
+import ru.serce.jnrfuse.ErrorCodes;
+import ru.serce.jnrfuse.FuseFillDir;
+import ru.serce.jnrfuse.FuseStubFS;
+import ru.serce.jnrfuse.struct.FileStat;
+import ru.serce.jnrfuse.struct.FuseFileInfo;
+
+import static jnr.constants.platform.OpenFlags.O_RDONLY;
+import static jnr.constants.platform.OpenFlags.O_WRONLY;
+
+/**
+ * Main FUSE implementation class.
+ *
+ * Implements the FUSE callbacks defined by jnr-fuse
+ */
+public final class TachyonFuseFs extends FuseStubFS {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+  private static final int MAX_OPEN_FILES = Integer.MAX_VALUE;
+  // Limits the number of translated (FUSE to TachyonURI)  paths that are kept
+  // in memory
+  private static final int MAX_CACHED_PATHS = 500;
+  private static final long[] UID_AND_GID = TachyonFuseUtils.getUidAndGid();
+
+  private final TachyonFileSystem mTFS;
+  private final Path mMountPoint;
+  private final Path mTachyonRootPath;
+  private final String mTachyonMaster;
+  // Keeps a cache of the most recently translated paths from String to TachyonURI
+  private final LoadingCache<String, TachyonURI> mPathResolverCache;
+
+  private final Object mOpenFilesLock;
+  // Table of open files with corresponding InputStreams and OutptuStreams
+  private final Map<Long, OpenFileEntry> mOpenFiles;
+  private long mOpenFileIds;
```

Comment:
maybe rename to `mNextOpenFileId`? `long mOpenFileIds` sounds like the number of open files. Please add javadoc to clarify what `mOpenFileIds` represents.